File: FE/web-app/src/modules/book/types/book.ts
import { BasePaging } from "../../shared";

export interface CategoryDTO
{
    name: string;
    thumbnail: File|null;
    parentId: number|null;
}

export interface CategoryListItemDTO
{
    id: number;
    name: string;
    thumbnail: string;
    parentId: number|null;
    parentName: number|null;
    createdAt: Date;
    updatedAt: Date;
}

export interface CategorySelectItemDTO
{
    id: number;
    name: string;
}

export interface BookDTO
{
    cover: File|null;
    name: string;
    author: string;
    publisher: string;
    publishYear: number;
    pageCount: number;
    language: string;
    price: number;
    stock: number;
    description: string;
    categoryIds: number[];
    images: {
        id: number|null,
        image: File|null
    }[]
}

export interface BookListItemDTO
{
    id: string;
    cover: string;
    name: string;
    price: number;
    currency: string;
    updatedAt: Date;
    avgRate: number;
    soldCount: number;
    stockCount: number;
    categories: string[];
}

export interface BoughtBookListItemDTO extends BookListItemDTO {
    purchaseCount: number;
    lastPurchasedAt: Date;
}

export interface BookToReviewListItemDTO 
{
    bookId: string;
    bookName: string;
    price: number;
    quantity: number;
    cover: string;
}

export interface BookReviewListItemDTO
{
    userId: string;
    userName: string;
    userAvatarUrl: string;
    createdAt: Date;
    rate: number;
    content: string;
    imageUrls: string[];
}

export interface CreateBookReviewDTO
{
    bookId: string;
    content: string;
    images: File[];
    rate: number;
}

export interface BookSearchDTO extends BasePaging
{
    name: string;
    minPrice: number|null;
    maxPrice: number|null;
    sortBy: string;
    order: string;
    categoryIds: number[];
}

export interface BookDetailDTO
{
    id: string;
    name: string;
    cover: string;
    author: string;
    categories: string[];
    images: {
        id: number,
        imageUrl: string
    }[];
    publisher: string;
    publishYear: number;
    pageCount: number;
    language: string;
    stockCount: number;
    currency: string;
    description: string;
    soldCount: number;
    price: number;
    avgRate: number;
    createdAt: Date;
    updatedAt: Date;
}

File: BE/Core/Entities/BookFeature/Book.cs
using Core.Entities.PaymentFeature;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Core.Entities.BookFeature
{
    public class Book
    {
        public Guid Id { get; set; }
        public string Cover { get; set; }
        public string Name { get; set; }
        public string Author { get; set; }
        public string Language { get; set; }
        public decimal Price { get; set; }
        public string Currency { get; set; }
        public decimal AvgRate { get; set; }
        public string Description { get; set; }
        public int SoldCount { get; set; }
        public int StockCount { get; set; }
        public int PageCount { get; set; }
        public string Publisher { get; set; }
        public int PublishYear { get; set; }
        public DateTime CreatedAt { get; set; }
        public DateTime UpdatedAt { get; set; }
        // Rela
        public List<Review> Reviews { get; set; }
        public List<BookCategory> BookCategories { get; set; }
        public List<OrderDetail> OrderDetails { get; set; }
        public List<CartItem> CartItems { get; set; }
        public List<BookImage> Images { get; set; }
    }
}
